data/transactions/verify/txn.go

```diff
@@ -0,0 +1,202 @@
+	if err := s.Txn.WellFormed(spec, proto); err != nil {
+		return err
+	}
+
+	zeroAddress := basics.Address{}
+	if s.Txn.Src() == zeroAddress {
+		return errors.New("empty address")
+	}
+
+	return stxnVerifyCore(s, &proto)
+}
+
+type asyncVerifyContext struct {
+	s     *transactions.SignedTxn
+	outCh chan error
+	proto *config.ConsensusParams
+}
+
+func stxnAsyncVerify(arg interface{}) interface{} {
+	cx := arg.(*asyncVerifyContext)
+	err := stxnVerifyCore(cx.s, cx.proto)
+	if err != nil {
+		cx.outCh <- err
+	} else {
+		close(cx.outCh)
+	}
+	return nil
+}
+
+func stxnVerifyCore(s *transactions.SignedTxn, proto *config.ConsensusParams) error {
+	numSigs := 0
+	hasSig := false
+	hasMsig := false
+	hasLogicSig := false
+	if s.Sig != (crypto.Signature{}) {
+		numSigs++
+		hasSig = true
+	}
+	if !s.Msig.Blank() {
+		numSigs++
+		hasMsig = true
+	}
+	if !s.Lsig.Blank() {
+		numSigs++
+		hasLogicSig = true
+	}
+	if numSigs == 0 {
+		return errors.New("signedtxn has no sig")
+	}
+	if numSigs > 1 {
+		return errors.New("signedtxn should only have one of Sig or Msig or LogicSig")
+	}
+
+	if hasSig {
+		if crypto.SignatureVerifier(s.Txn.Src()).Verify(s.Txn, s.Sig) {
+			return nil
+		}
+		return errors.New("signature validation failed")
+	}
+	if hasMsig {
+		if ok, _ := crypto.MultisigVerify(s.Txn, crypto.Digest(s.Txn.Src()), s.Msig); ok {
+			return nil
+		}
+		return errors.New("multisig validation failed")
+	}
+	if hasLogicSig {
+		return LogicSig(&s.Lsig, proto, s)
+	}
+	return errors.New("has one mystery sig. WAT?")
+}
+
+// LogicSig checks that the signature is valid and that the program is basically well formed.
+// It does not evaluate the logic.
+func LogicSig(lsig *transactions.LogicSig, proto *config.ConsensusParams, stxn *transactions.SignedTxn) error {
+	if len(lsig.Logic) == 0 {
+		return errors.New("LogicSig.Logic empty")
+	}
+	version, vlen := binary.Uvarint(lsig.Logic)
+	if vlen <= 0 {
+		return errors.New("LogicSig.Logic bad version")
+	}
+	if version > proto.LogicSigVersion {
+		return errors.New("LogicSig.Logic version too new")
+	}
+	if uint64(lsig.Len()) > proto.LogicSigMaxSize {
+		return errors.New("LogicSig.Logic too long")
+	}
+
+	ep := logic.EvalParams{Txn: stxn, Proto: proto}
+	cost, err := logic.Check(lsig.Logic, ep)
+	if err != nil {
+		return err
+	}
+	if cost > int(proto.LogicSigMaxCost) {
+		return fmt.Errorf("LogicSig.Logic too slow, %d > %d", cost, proto.LogicSigMaxCost)
+	}
+
+	hasSig := false
+	hasMsig := false
+	numSigs := 0
+	if lsig.Sig != (crypto.Signature{}) {
+		hasSig = true
+		numSigs++
+	}
+	if !lsig.Msig.Blank() {
+		hasMsig = true
+		numSigs++
+	}
+	if numSigs == 0 {
+		// if the txn.Sender == hash(Logic) then this is a (potentially) valid operation on a contract-only account
+		program := transactions.Program(lsig.Logic)
+		lhash := crypto.HashObj(&program)
+		if crypto.Digest(stxn.Txn.Sender) == lhash {
```

Comment:
nit: we use `Src()` below instead of `Sender`, though I just checked and they're the same.